src/main/java/org/eclipse/uprotocol/mqtt/HiveMqMQTT5Client.java

```diff
@@ -0,0 +1,280 @@
+    private static final String USER_PROPERTIES_KEY_FOR_TOKEN = "10";
+    private static final String USER_PROPERTIES_KEY_FOR_TRACEPARENT = "11";
+    private static final String USER_PROPERTIES_KEY_FOR_PAYLOAD_FORMAT = "12";
+    private static final String USER_PROPERTIES_KEY_FOR_SOURCE_NAME = "3";
+    private static final String USER_PROPERTIES_KEY_FOR_SINK_NAME = "4";
+    private final Mqtt5AsyncClient client;
+    private final UUri source;
+
+    public HiveMqMQTT5Client(UUri source, Mqtt5Client client) {
+        this.client = client.toAsync();
+        this.source = source;
+    }
+
+    @Override
+    public UStatus send(UMessage uMessage) {
+        return sendAsync(uMessage).join();
+    }
+
+    public CompletableFuture<UStatus> sendAsync(UMessage uMessage) {
+        LOG.trace("should send a message:\n{}", uMessage);
+        CompletableFuture<UStatus> result = new CompletableFuture<>();
+
+        Mqtt5UserProperties userProperties = Mqtt5UserProperties.builder()
+                .add("0", "1")
+                .add(USER_PROPERTIES_KEY_FOR_ID, UuidSerializer.serialize(uMessage.getAttributes().getId()))
+                .add(USER_PROPERTIES_KEY_FOR_MESSAGE_TYPE, Integer.toString(uMessage.getAttributes().getTypeValue()))
+                .add(USER_PROPERTIES_KEY_FOR_SOURCE_NAME, uMessage.getAttributes().getSource().getAuthorityName())
+                .add(USER_PROPERTIES_KEY_FOR_SINK_NAME, uMessage.getAttributes().getSink().getAuthorityName())
+                .add(USER_PROPERTIES_KEY_FOR_PRIORITY, Integer.toString(uMessage.getAttributes().getPriorityValue()))
+                .add(USER_PROPERTIES_KEY_FOR_TTL, Integer.toString(uMessage.getAttributes().getTtl()))
+                .add(USER_PROPERTIES_KEY_FOR_PERMISSION_LEVEL, Integer.toString(uMessage.getAttributes().getPermissionLevel()))
+                .add(USER_PROPERTIES_KEY_FOR_COMMSTATUS, Integer.toString(uMessage.getAttributes().getCommstatusValue()))
+                .add(USER_PROPERTIES_KEY_FOR_REQID, UuidSerializer.serialize(uMessage.getAttributes().getReqid()))
+                .add(USER_PROPERTIES_KEY_FOR_TOKEN, uMessage.getAttributes().getToken())
+                .add(USER_PROPERTIES_KEY_FOR_TRACEPARENT, uMessage.getAttributes().getTraceparent())
+                .add(USER_PROPERTIES_KEY_FOR_PAYLOAD_FORMAT, Integer.toString(uMessage.getAttributes().getPayloadFormatValue()))
+                .build();
+
+        Mqtt5PublishBuilder.Send.Complete<CompletableFuture<Mqtt5PublishResult>> sendHandle = client.publishWith()
+                .topic(getTopicForSending(uMessage.getAttributes().getSource(), uMessage.getAttributes().getSink()))
+                .payload(uMessage.getPayload().toByteArray())
+                .userProperties(userProperties);
+
+        if (uMessage.getAttributes().hasTtl()) {
+            sendHandle.messageExpiryInterval(Math.round(uMessage.getAttributes().getTtl() / 1000d));
+        }
+
+        sendHandle
```

Comment:
Ok I see what you mean by the send async vs send() sync however we need to know that the message was sent correctly and if we start adding async(), it will change the behavior of the transport implementations. Lets discuss tomorrow.

Reply:
as async is part of up-java 0.1.11, I think, this is already discussed